### ISSUE TITLE
Color space discrepancy on video

### ISSUE THREAD

**Issue by julianpalladino** [NVIDIA/DALI#4441]
Hi!
First of all, I would like to thank you for this amazing library, not only for making it free and available for everyone but also for making it open source.

I'm having an issue with the color spaces when using `nvidia.dali.fn.readers.video` and I was hoping to get feedback from you. I'm not sure if this is a bug or if I'm doing something wrong.

Please look at the following minimal example in which I can replicate the error.

Given an image `test_img.jpg`, we can copy it 30 times and create a video using those 30 images as frames using `FFMPEG`, with the following script:

```
import os
import subprocess
from PIL import Image
import shutil

def create_video():
    number_of_frames = 30
    # We copy `test_img.jpg` to 00.jpg, 01.jpg, 02.jpg, ..., 29.jpg
    for i in range(number_of_frames):
        shutil.copyfile('test_img.jpg', '%02d.jpg' % i)

    # We use 0.275 bpp and the image size to determine the frame rate 
    img = Image.open('test_img.jpg')
    rate = round(0.275 * img.size[0] * img.size[1] * 25 / 1e6, 1)
    cmd = 'ffmpeg -r 25 -i %02d.jpg -c:v libx265 -x265-params keyint=1:min-keyint=1 -hide_banner' + \
        f' -b:v {rate} -maxrate {rate} -y -colorspace bt470bg -color_primaries bt470bg ' + \
        '-color_range 2 -movflags +faststart video.mp4'
    subprocess.run(cmd.split(' '))

create_video()
```

Then, we can load the video using `fn.readers.video` and save the first frame:

```
import os
import torch
torch.backends.cudnn.benchmark = True
import numpy as np
from PIL import Image
from nvidia import dali
from nvidia.dali import pipeline_def, Pipeline
import nvidia.dali.fn as fn
import nvidia.dali.types as types
from nvidia.dali.plugin.pytorch import DALIGenericIterator
from nvidia.dali.plugin.pytorch import feed_ndarray

@pipeline_def
def generate_video_pipeline():
    videos, labels, frame_num  = fn.readers.video(
        device="gpu",
        filenames=['video.mp4'],
        sequence_length=8,
        shard_id=0,
        num_shards=1,
        random_shuffle=False,
        initial_fill=10,
        name='Reader', 
        normalized=False,
        dtype=types.UINT8,
        enable_frame_num=True,
        pad_sequences=True,
        labels=[],
        read_ahead=True)
    return videos, frame_num

video_pipeline = generate_video_pipeline(
    batch_size=1, # "Batches" consist of multiple "sequences". We want only 1
    num_threads=2,
    device_id=0)

video_pipeline.build()

video_pipeline.schedule_run()
pipe_out_frame, _ = video_pipeline.share_outputs()

current_frame_dali = pipe_out_frame.as_tensor()
current_frame_dali.squeeze()

dev_id = video_pipeline.device_id
torch_gpu_device = torch.device('cuda', dev_id)

current_batch_torch = torch.empty(
                current_frame_dali.shape(),
                dtype=torch.uint8,
                device=torch_gpu_device)

stream = torch.cuda.current_stream(device=dev_id)
dali.plugin.pytorch.feed_ndarray(
    current_frame_dali, current_batch_torch, cuda_stream=stream)

video_pipeline.release_outputs()

Image.fromarray((np.array(current_batch_torch[0,:,:,:].cpu()).astype(np.uint8))).save('first_frame.png')
```

However, the frame obtained with DALI differs from the original image. Here's an example:

**Original image**
![test_img](https://user-images.githubusercontent.com/8797947/201389107-19ec869a-da6e-4beb-8f5c-1fedfc0f807a.jpg)

**First frame from DALI**
![image](https://user-images.githubusercontent.com/8797947/201392030-d236292d-8016-4946-815c-2e41ae3eae8a.png)
At first glance, the frame may look correct. However, when looking at both of them side by side there seems to be a difference, maybe in the color range.

Here's a histogram comparison between the original image and DALI's frame:
![hist_comparison](https://user-images.githubusercontent.com/8797947/201395853-f1598e5f-e42c-4caa-8321-da012f6a89bf.png)

Is there a way for me to specify to DALI the color range I want to use? If there isn't, What is the actual color space used by the DALI decoder?

Thank you very much in advance, and if there's anything else I can provide to help find the answers to my questions please let me know.

**Comment by julianpalladino** [comment]
Hi @JanuszL ,
Thank you very much for your quick reply. I've tried `ffmpeg -i input.mp4 %04d.jpg` and was getting the same kind of incorrect decoding than what I was getting with DALI.

So I added the following into the video creation FFmpeg command: 
```
keyint=1:min-keyint=1:range=full -dst_range 1 -pix_fmt yuvj420p
```

Making the full command:
```
cmd = f'ffmpeg -r 25 -i %02d.jpg -c:v libx265 -x265-params keyint=1:min-keyint=1:range=full -dst_range 1 
-pix_fmt yuvj420p -hide_banner -b:v {rate}M -maxrate {rate}M -y -colorspace bt470bg -color_primaries bt470bg 
-color_range 2 -movflags +faststart {output_video}'
```
And now I'm getting the correct images from `ffmpeg -i input.mp4 %04d.jpg`. However, DALI is still decoding it incorrectly. Here's a sample:

**Original image**
![image](https://user-images.githubusercontent.com/8797947/201730385-14f0a899-b019-43fb-b764-df11de05ec4f.png)

**Video frame obtained with FFmpeg**
![image](https://user-images.githubusercontent.com/8797947/201730299-e383f6b7-d789-44c1-906b-5180198ea0ae.png)

**Video frame obtained with DALI**
![image](https://user-images.githubusercontent.com/8797947/201731077-2625bce6-2b33-4e63-8308-2d092afa9e02.png)

Does DALI parse the VUI headers? We can't seem to find it in the code.

**Comment by JanuszL** [comment]
@julianpalladino,

Thank you for the detailed repro steps. Let me try this out and I will get back to you with more info.

**Comment by julianpalladino** [comment]
Great! If there's anything I can help you with, please let me know. Thanks!

**Comment by JanuszL** [comment]
@julianpalladino,

DALI assumes that the video uses BT.601 color format. In your case you use bt470bg.
Let me see if and how we can extend DALI to support different color video formats.

**Comment by mtnwrw** [comment]
The color matrices for BT.601 and BT.470 are the same. The issue that Julian is experiencing is related to the color range (a.k.a.  "swing").  The input data in Julian's case is encoded as "full swing" data, whereas it seems that Dali assumes it is "studio swing", this explains the discrepancy.

**Comment by JanuszL** [comment]
@mtnwrw,

True that. I'm not that experience with video internals. Now I see that `yuvj420p` states the full range (unless I'm wrong again).

**Comment by mtnwrw** [comment]
I took the liberty to look into your test-videos and I think you might have a small data problem. All your test videos are encoded in studio swing (i.e. not the full range). Aside from the fact that you are effectively discarding about 14% of available information as opposed to a full swing representation, there is more trouble down the road.

When using a network that was trained using JPEGs or PNGs, using DALI as drop-in replacement for image loading from videos instead of JPEGs during inference time will most likely lower the accuracy of that network, because of the diminished range. Worst case even adding unwanted luminance/color scaling. I assume that is how Julian stumbled across this issue.

See [this article](https://en.wikipedia.org/wiki/YUV#Studio_swing_for_YCbCr_BT.601) for an explanation of studio vs. full swing.

**Comment by jpalladino-fy** [comment]
Hi @JanuszL,
I tested the changes with DALI's nightly build and it's working now! 🎉 Thank you very much for the fix.

**Comment by JanuszL** [comment]
It should be a part of 1.20 which has just been released.
